File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int avgNodeCount = 0;
    pair<int,int> averageOfSubtree_01(TreeNode* node) {
        if(!node->left && !node->right){
            avgNodeCount ++;
            return {node->val,1};
        }
        
        pair leftPair = {0,0};
        pair rightPair = {0,0};
        pair myPair = {0,0};
        
        if(node->left) leftPair = averageOfSubtree_01(node->left);
        
        if(node->right) rightPair = averageOfSubtree_01(node->right);
        
        myPair.first = leftPair.first + rightPair.first + node->val;
        myPair.second = leftPair.second + rightPair.second + 1;
        
        if(myPair.first / myPair.second == node->val) avgNodeCount++;
        
        return myPair;
    }
    int averageOfSubtree(TreeNode* root) {
        averageOfSubtree_01(root);
        return avgNodeCount;
    }
};

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public class pair {
        int sum = 0;
        int nodeCount=0;
        pair(int sum,int nodeCount) {
            this.sum = sum;
            this.nodeCount = nodeCount;
        }
        pair(){}
    }
    int avgNodeCount=0;
    public pair averageOfSubtree_01(TreeNode node) {
        if(node.left ==null && node.right == null) {
            avgNodeCount++;
            return new pair(node.val,1);
        }
        
        pair leftP = new pair();
        pair rightP = new pair();
        pair myPair = new pair();
        
        if(node.left != null) leftP = averageOfSubtree_01(node.left);
        if(node.right!= null) rightP = averageOfSubtree_01(node.right);
        
        myPair.sum = leftP.sum + rightP.sum + node.val;
        myPair.nodeCount = leftP.nodeCount + rightP.nodeCount +1;
        
        if(myPair.sum / myPair.nodeCount == node.val) avgNodeCount++;
        
        return myPair;
    }
    
    public int averageOfSubtree(TreeNode root) {
        averageOfSubtree_01(root);
        return avgNodeCount;
    }
}